src/source/geojson_source_diff.ts

```
@@ -0,0 +1,123 @@
+
+export type GeoJSONFeatureId = number | string;
+export interface GeoJSONSourceDiff {
+    removeAll?: boolean;
+    removed?: Array<GeoJSONFeatureId>;
+    add?: Array<GeoJSON.Feature<DirectGeometry> & {id: GeoJSONFeatureId}>;
+    update?: Array<GeoJSONFeatureDiff>;
+}
+
+export interface GeoJSONFeatureDiff {
+    id: GeoJSONFeatureId;
+    newGeometry?: DirectGeometry;
+    removeAllProperties?: boolean;
+    removeProperties?: Array<string>;
+    addOrUpdateProperties?: Array<{key: string; value: any}>;
+}
+
+export type DirectGeometry = GeoJSON.Point | GeoJSON.MultiPoint | GeoJSON.LineString | GeoJSON.MultiLineString | GeoJSON.Polygon | GeoJSON.MultiPolygon;
+export type UpdateableGeoJSON = GeoJSON.Feature<DirectGeometry> & {id: GeoJSONFeatureId} | GeoJSON.FeatureCollection<DirectGeometry> & {features: {id: GeoJSONFeatureId}[]};
+
+export function isDirectGeometry(geometry: GeoJSON.Geometry): geometry is DirectGeometry {
+    return geometry.type !== 'GeometryCollection';
+}
+
+export function isUpdateable(data: string | GeoJSON.GeoJSON): data is UpdateableGeoJSON {
+    if (typeof data === 'string') {
+        return false;
+    }
+
+    if (data.type === 'Feature' && data.id != null && isDirectGeometry(data.geometry)) {
+        return true;
+    }
+
+    if (data.type === 'FeatureCollection' && data.features.every(feature => feature.id != null && isDirectGeometry(feature.geometry))) {
```

this doesn't insulate against duplicate IDs does it?  And it might be rather expensive to actually do the duplicates check...

---

The applySourceDiff will implicitly chose the final feature with a given ID

---

right, however `isUpdateable` will return true for `{"features": [{"id":"a", ...}, {"id":"a", ...}]` and then `getUpdateable` will silently drop the first feature.  I'm pretty sure this changes the current behavior of setData, which isn't a good idea.

---

That's a good point. We don't technically break the setData behavior, but passing the exact same features through the `updateData` call will result in fewer visible features.

---

resolved by https://github.com/mfedderly/maplibre-gl-js/blob/geojson-diff-to-worker/src/source/geojson_source_diff.ts#L49-L53